File: config-auth-core-full-kt/assets/auth-core-full-template-kt/src/main/kotlin/com/example/auth/role/entity/Role.kt
package com.example.auth.role.entity

import com.example.shared.domain.vo.Id
import com.example.shared.domain.vo.Name

data class Role private constructor(
    val id: Id,
    val name: Name,
    val description: String? = null
) {
    companion object {
        fun create(id: String? = null, name: String, description: String? = null): Role =
            tryCreate(id, name, description).getOrThrow()

        fun tryCreate(id: String? = null, name: String, description: String? = null): Result<Role> {
            val validId = Id.tryCreate(id).getOrElse { return Result.failure(it) }
            val validName = Name.tryCreate(name).getOrElse { return Result.failure(it) }
            return Result.success(Role(validId, validName, description))
        }
    }

    override fun equals(other: Any?): Boolean = other is Role && other.id == this.id
    override fun hashCode(): Int = id.hashCode()
}

File: config-auth-core-full-cs/assets/auth-core-full-template-cs/Domain/Entities/Role.cs
using Project.Shared.Kernel.Domain.Base;
using Project.Shared.Kernel.Domain.Results;

namespace Project.Auth.Domain.Entities;

public class Role : Entity
{
    public string Name { get; private set; } = string.Empty;
    private readonly List<Permission> _permissions = new();
    public IReadOnlyCollection<Permission> Permissions => _permissions.AsReadOnly();

    private Role() { }

    public static Result<Role> Create(string name)
    {
        if (string.IsNullOrWhiteSpace(name))
            return Result<Role>.Failure("Role name is required.");

        return Result<Role>.Success(new Role { Id = Guid.NewGuid(), Name = name.Trim() });
    }

    public Result AddPermission(Permission permission)
    {
        if (_permissions.Any(p => p.Key == permission.Key))
            return Result.Failure($"Permission '{permission.Key}' already assigned.");

        _permissions.Add(permission);
        return Result.Success();
    }

    public void RemovePermission(string permissionKey)
    {
        _permissions.RemoveAll(p => p.Key == permissionKey);
    }

    public bool HasPermission(string permissionKey)
    {
        return _permissions.Any(p => p.Key == permissionKey);
    }
}
